include/RAJA/exec-cuda/reduce_cuda.hxx

```diff
@@ -830,154 +1175,221 @@ private:
+   * \brief Operator that returns reduced sum value.
+   *
+   * Note: accessor only executes on host.
+   */
+  T get() { return operator T(); }
 
-  int m_myID;
+  /*!
+   * \brief Operator that adds value to sum.
+   *
+   * Note: only operates on device.
+   */
+  __device__ ReduceSum<cuda_reduce_atomic<BLOCK_SIZE, Async>, T> const &
+  operator+=(T val) const
+  {
+    extern __shared__ unsigned char sd_block[];
+    T *sd = reinterpret_cast<T *>(&sd_block[m_smem_offset]);
 
-  T m_init_val;
-  T m_reduced_val;
+    int threadId = threadIdx.x + blockDim.x * threadIdx.y
+                   + (blockDim.x * blockDim.y) * threadIdx.z;
 
-  CudaReductionBlockTallyType *m_tallydata;
+    sd[threadId] += val;
 
-  // Sanity checks for block size
+    return *this;
+  }
+
+private:
+  /*!
+   * \brief Default constructor is declared private and not implemented.
+   */
+  ReduceSum<cuda_reduce_atomic<BLOCK_SIZE, Async>, T>();
+
+  /*!
+   * \brief Pointer to host tally block cache slot for this reduction variable.
+   */
+  CudaReductionTallyTypeAtomic<T> *m_tally_host = nullptr;
+
+  /*!
+   * \brief Pointer to device tally block slot for this reduction variable.
+   */
+  CudaReductionTallyTypeAtomic<T> *m_tally_device = nullptr;
+
+  /*!
+   * \brief My cuda reduction variable ID.
+   */
+  int m_myID = -1;
+
+  /*!
+   * \brief Byte offset into dynamic shared memory.
+   */
+  int m_smem_offset = -1;
+
+  /*!
+   * \brief If this variable is a copy or not; only original may release memory 
+   *        or perform finalization.
+   */
+  bool m_is_copy_host = false;
+  bool m_is_copy_device = false;
+  bool m_finish_reduction = false;
+
+  // Sanity checks for block size and template type size
   static constexpr bool powerOfTwoCheck = (!(BLOCK_SIZE & (BLOCK_SIZE - 1)));
   static constexpr bool reasonableRangeCheck =
       ((BLOCK_SIZE >= 32) && (BLOCK_SIZE <= 1024));
+  static constexpr bool sizeofcheck =
+      ((sizeof(T) <= sizeof(CudaReductionDummyDataType))
+       && (sizeof(CudaReductionTallyType<T>)
+           <= sizeof(CudaReductionDummyTallyType))
+       && (sizeof(CudaReductionBlockType<T>)
+           <= sizeof(CudaReductionDummyBlockType)));
   static_assert(powerOfTwoCheck, "Error: block sizes must be a power of 2");
   static_assert(reasonableRangeCheck,
                 "Error: block sizes must be between 32 and 1024");
+  static_assert(sizeofcheck,
+      "Error: type must be of size <= " 
+      RAJA_STRINGIFY_MACRO(RAJA_CUDA_REDUCE_VAR_MAXSIZE));
 };
 
-///
-/// Each ReduceMinLoc or ReduceMaxLoc object uses retiredBlocks as a way
-/// to complete the reduction in a single pass. Although the algorithm
-/// updates retiredBlocks via an atomicAdd(int) the actual reduction values
-/// do not use atomics and require a finishing stage performed
-/// by the last block.
-///
-__device__ __managed__ unsigned int retiredBlocks[RAJA_MAX_REDUCE_VARS];
+
+// set index first to avoid changes to val1 or val2 when writing to val_set
```

Comment:
These macros unfortunately are specific to the cuda version. The cuda version uses struct of arrays in shared memory to avoid bank conflicts so it can't use the generic max/minloc macros.